File: production/client/src/main/scala/kpn/client/components/subset/routes/SubsetOrphanRouteFilter.scala
// TODO migrate to Angular
package kpn.client.components.subset.routes

import japgolly.scalajs.react.CallbackTo
import kpn.client.filter.BooleanFilter
import kpn.client.filter.FilterOptions
import kpn.client.filter.Filters
import kpn.client.filter.TimeFilterKind
import kpn.client.filter.TimestampFilter
import kpn.shared.RouteSummary
import kpn.shared.TimeInfo

class SubsetOrphanRouteFilter(timeInfo: TimeInfo, criteria: SubsetOrphanRouteFilterCriteria, updateCriteria: (SubsetOrphanRouteFilterCriteria) => Unit) {

  private val definedInNetworkRelationFilter = new BooleanFilter[RouteSummary](
    "definedInNetworkRelation",
    criteria.broken,
    _.isBroken,
    CallbackTo {
      updateCriteria(criteria.copy(broken = None))
    },
    CallbackTo {
      updateCriteria(criteria.copy(broken = Some(true)))
    },
    CallbackTo {
      updateCriteria(criteria.copy(broken = Some(false)))
    }
  )

  private val lastUpdatedFilter = new TimestampFilter[RouteSummary](
    criteria.lastUpdated,
    _.timestamp,
    timeInfo,
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.ALL))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.LAST_WEEK))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.LAST_MONTH))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.LAST_YEAR))
    },
    CallbackTo {
      updateCriteria(criteria.copy(lastUpdated = TimeFilterKind.OLDER))
    }
  )

  private val allFilters = new Filters[RouteSummary](
    lastUpdatedFilter
  )

  def filter(routes: Seq[RouteSummary]): Seq[RouteSummary] = {
    routes.filter(allFilters.passes)
  }

  def filterOptions(routes: Seq[RouteSummary]): FilterOptions = {

    val totalCount = routes.size
    val filteredCount = routes.count(allFilters.passes)

    val lastUpdated = lastUpdatedFilter.filterOptions(allFilters, routes)

    val groups = Seq(
      lastUpdated
    ).flatten

    FilterOptions(
      filteredCount,
      totalCount,
      groups: _*
    )
  }
}

File: frontend/src/app/analysis/subset/internal/orphan-routes/components/subset-orphan-route-filter.ts
import { WritableSignal } from '@angular/core';
import { OrphanRouteInfo } from '@api/common/orphan-route-info';
import { TimeInfo } from '@api/common/time-info';
import { BooleanFilter } from '@app/shared/kpn/filter/boolean-filter';
import { FilterOptions } from '@app/shared/kpn/filter/filter-options';
import { Filters } from '@app/shared/kpn/filter/filters';
import { TimestampFilter } from '@app/shared/kpn/filter/timestamp-filter';
import { TimestampFilterKind } from '@app/shared/kpn/filter/timestamp-filter-kind';
import { SubsetOrphanRouteFilterCriteria } from './subset-orphan-route-filter-criteria';

export class SubsetOrphanRouteFilter {
  private readonly brokenFilter = new BooleanFilter<OrphanRouteInfo>(
    'investigate',
    this.criteria().broken,
    (row) => row.isBroken,
    () => this.criteria.update((c) => ({ ...c, broken: null })),
    () => this.criteria.update((c) => ({ ...c, broken: true })),
    () => this.criteria.update((c) => ({ ...c, broken: false }))
  );
  private readonly lastUpdatedFilter = new TimestampFilter<OrphanRouteInfo>(
    this.criteria().lastUpdated,
    (row) => row.lastUpdated,
    this.timeInfo,
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.all })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.lastWeek })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.lastMonth })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.lastYear })),
    () => this.criteria.update((c) => ({ ...c, lastUpdated: TimestampFilterKind.older }))
  );
  private readonly allFilters = new Filters<OrphanRouteInfo>(
    this.brokenFilter,
    this.lastUpdatedFilter
  );

  constructor(
    private readonly criteria: WritableSignal<SubsetOrphanRouteFilterCriteria>,
    private readonly timeInfo: TimeInfo
  ) {}

  filter(routes: OrphanRouteInfo[]): OrphanRouteInfo[] {
    return routes.filter((route) => this.allFilters.passes(route));
  }

  filterOptions(routes: OrphanRouteInfo[]): FilterOptions {
    const totalCount = routes.length;
    const filteredCount = routes.filter((route) => this.allFilters.passes(route)).length;

    const broken = this.brokenFilter.filterOptions(this.allFilters, routes);
    const lastUpdated = this.lastUpdatedFilter.filterOptions(this.allFilters, routes);

    const groups = [broken, lastUpdated].filter((g) => g !== null);

    return new FilterOptions(filteredCount, totalCount, groups);
  }
}
